dask/array/backends.py

```diff
@@ -203,3 +212,61 @@ def _tensordot_scipy_sparse(a, b, axes):
         return a * b
     elif a_axis == 1 and b_axis == 1:
         return a * b.T
+
+
+@numel_lookup.register(np.ma.masked_array)
+def _numel_masked(x, **kwargs):
+    return chunk.sum(np.ones_like(x), **kwargs)
+
+
+@numel_lookup.register((object, np.ndarray))
+def _numel_ndarray(x, **kwargs):
+    return _numel(x, coerce_ndarray=True, **kwargs)
+
+
+def _numel_arraylike(x, **kwargs):
+    return _numel(x, coerce_ndarray=False, **kwargs)
+
+
+def _numel(x, coerce_ndarray: bool, **kwargs):
```

Comment:
This is mostly the same as the existing `numel` function, but has an additional argument in `coerce_ndarray`, which, if true, forces the result to be an `ndarray`. If false, it lets `np.full_like` determine the output type, which will usually be the same as what we give it. We want `coerce_ndarray` for sparse arrays, and we don't want it for masked arrays, cupy, etc.

This is all private, so I think it's fairly safe (though I suppose I could obfuscate the kwarg name a bit more)